presto-main/src/main/java/com/facebook/presto/operator/LookupJoinOperator.java

```diff
@@ -136,94 +196,371 @@ public boolean needsInput()
+            /*
+             * We do not have input probe and we won't have any, as we're finishing.
+             * Let LookupSourceFactory know LookupSources can be disposed as far as we're concerned.
+             */
+            finishRegularInput();
+            unspilling = true;
+        }
+
+        if (probe == null && unspilling && !finished) {
+            /*
+             * If no current partition or it was exhausted, unspill next one.
+             * Add input there when it needs one, produce output. Be Happy.
+             */
+            if (!tryUnspillNext()) {
+                return null;
+            }
+        }
 
-        try (LookupSourceProvider.LookupSourceLease lookupSourceLease = lookupSourceProvider.leaseLookupSource()) {
-            return getOutput(lookupSourceLease.getLookupSource());
+        if (probe != null) {
+            processProbe();
         }
+
+        return producePage();
     }
 
-    private Page getOutput(LookupSource lookupSource)
+    private void finishRegularInput()
     {
-        // join probe page with the lookup source
-        Counter lookupPositionsConsidered = new Counter();
-        if (probe != null) {
-            while (true) {
-                if (probe.getPosition() >= 0) {
-                    if (!joinCurrentPosition(lookupSource, lookupPositionsConsidered)) {
-                        break;
-                    }
-                    if (!currentProbePositionProducedRow) {
-                        currentProbePositionProducedRow = true;
-                        if (!outerJoinCurrentPosition(lookupSource)) {
-                            break;
-                        }
-                    }
-                }
-                currentProbePositionProducedRow = false;
-                if (!advanceProbePosition(lookupSource)) {
-                    break;
-                }
-                statisticsCounter.recordProbe(joinSourcePositions);
-                joinSourcePositions = 0;
+        checkState(partitionedConsumption == null);
+        partitionedConsumption = lookupSourceFactory.finishProbeOperator(lookupJoinsCount);
+    }
+
+    private boolean tryUnspillNext()
+    {
+        verify(probe == null);
+
+        if (!partitionedConsumption.isDone()) {
+            return false;
+        }
+        if (lookupPartitions == null) {
+            lookupPartitions = getDone(partitionedConsumption).beginConsumption();
+        }
+
+        if (unspilledInputPages.hasNext()) {
+            addInput(unspilledInputPages.next());
+            verify(probe != null);
+            return true;
+        }
+        else if (unspilledLookupSource.isPresent()) {
+            if (!unspilledLookupSource.get().isDone()) {
+                // Not unspilled yet
+                return false;
+            }
+            LookupSource lookupSource = getDone(unspilledLookupSource.get()).get();
+            unspilledLookupSource = Optional.empty();
+
+            lookupSourceProvider.close();
```

Comment:
Why are we closing `lookupSourceProvider` here? N times for every unspilled partition.
If there is no better place at lease provide comment with explanation.